File: rest-common/src/main/java/org/chronopolis/rest/entities/AceToken.java
package org.chronopolis.rest.entities;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import java.util.Date;

/**
 * Entity representing an ACE Token in our DB
 *
 * Created by shake on 2/4/15.
 */
@Entity
public class AceToken extends PersistableEntity {

    @JoinColumn(name = "bag")
    @ManyToOne(fetch = FetchType.LAZY)
    private Bag bag;

    private Date createDate;
    private String filename;
    private String proof;
    private String imsHost;
    private String imsService;
    private String algorithm;
    private Long round;

    protected AceToken() { // JPA
    }

    public AceToken(final Bag bag,
                    final Date createDate,
                    final String filename,
                    final String proof,
                    final String imsHost,
                    final String imsService,
                    final String algorithm,
                    final Long round) {
        this.bag = bag;
        this.createDate = createDate;
        this.filename = filename;
        this.proof = proof;
        this.imsHost = imsHost;
        this.imsService = imsService;
        this.algorithm = algorithm;
        this.round = round;
    }

    public Bag getBag() {
        return bag;
    }

    public void setBag(final Bag bag) {
        this.bag = bag;
    }

    public Date getCreateDate() {
        return createDate;
    }

    public void setCreateDate(final Date createDate) {
        this.createDate = createDate;
    }

    public String getFilename() {
        return filename;
    }

    public void setFilename(final String filename) {
        this.filename = filename;
    }

    public String getProof() {
        return proof;
    }

    public void setProof(final String proof) {
        this.proof = proof;
    }

    public String getImsService() {
        return imsService;
    }

    public void setImsService(final String imsService) {
        this.imsService = imsService;
    }

    public String getAlgorithm() {
        return algorithm;
    }

    public void setAlgorithm(final String algorithm) {
        this.algorithm = algorithm;
    }

    public Long getRound() {
        return round;
    }

    public void setRound(final Long round) {
        this.round = round;
    }

    public String getImsHost() {
        return imsHost;
    }

    public AceToken setImsHost(String imsHost) {
        this.imsHost = imsHost;
        return this;
    }
}

File: rest-entities/src/main/kotlin/org/chronopolis/rest/entities/AceToken.kt
package org.chronopolis.rest.entities

import java.util.Date
import javax.persistence.Entity
import javax.persistence.FetchType
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToOne

/**
 * AceToken connected to a single [BagFile]
 *
 * @property proof the proof returned by the IMS
 * @property round the round returned by the IMS
 * @property imsService the digest service used by the IMS
 * @property algorithm the digest algorithm used by the IMS
 * @property imsHost the hostname of the IMS
 * @property createDate the create date of the Ace Token returned by the IMS
 * @property bag the corresponding [Bag]
 * @property file the corresponding [BagFile]
 *
 * @author shake
 */
@Entity
class AceToken(
        var proof: String = "",
        var round: Long = -1,
        var imsService: String = "",
        var algorithm: String = "",
        var imsHost: String = "",
        var createDate: Date = Date(),

        @ManyToOne(fetch = FetchType.LAZY)
        var bag: Bag = Bag(),

        @OneToOne
        @JoinColumn(name = "file_id")
        var file: BagFile = BagFile()
) : PersistableEntity() {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AceToken

        if (proof != other.proof) return false
        if (round != other.round) return false
        if (imsService != other.imsService) return false
        if (algorithm != other.algorithm) return false
        if (imsHost != other.imsHost) return false
        if (createDate != other.createDate) return false

        return true
    }

    override fun hashCode(): Int {
        var result = proof.hashCode()
        result = 31 * result + round.hashCode()
        result = 31 * result + imsService.hashCode()
        result = 31 * result + algorithm.hashCode()
        result = 31 * result + imsHost.hashCode()
        result = 31 * result + createDate.hashCode()
        return result
    }

}